### ISSUE TITLE
Unexpected behaviour on timer cancel (broken) ?

### ISSUE THREAD

**Issue by tonythomas01** [jruizgit/rules#202]
We were observing this on our production instance, and here is a simple test for the same. 
```
import logging
import time

import redis
from durable.lang import *

r = redis.Redis()
r.flushdb()

with ruleset('reproduce_timer_issue'):
    @when_all(m.action == 'start_timer')
    def start_timer(c):
        logging.info("Starting timer for sid: {sid}".format(sid=c.m.sid))
        c.start_timer("important_timer", 5)

    @when_all(m.action == 'cancel_timer')
    def cancel_timer(c):
        logging.info("Cancelling timer for sid: {sid}".format(sid=c.m.sid))
        c.cancel_timer("important_timer")

    @when_all(timeout("important_timer"))
    def timer_fired(c):
        logging.info("Important timer with sid {sid} expired.".format(sid=c.m.sid))

    @when_start
    def on_start(host):
        for id in range(10):
            host.post('reproduce_timer_issue', { 'action': 'start_timer', "sid": id })
        host.post('reproduce_timer_issue', {'action': 'cancel_timer', "sid": 5})

logging.getLogger().setLevel(logging.INFO)
run_all()

Produces output:
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
INFO:root:Starting timer for sid: 0
INFO:root:Starting timer for sid: 1
INFO:root:Starting timer for sid: 2
INFO:root:Starting timer for sid: 3
INFO:root:Starting timer for sid: 4
INFO:root:Starting timer for sid: 5
INFO:root:Cancelling timer for sid: 5
INFO:root:Starting timer for sid: 6
INFO:root:Starting timer for sid: 7
INFO:root:Starting timer for sid: 8
INFO:root:Starting timer for sid: 9
INFO:root:Important timer with sid 6 expired.
INFO:root:Important timer with sid 7 expired.
INFO:root:Important timer with sid 8 expired.
INFO:root:Important timer with sid 9 expired.
``` 

as you can see - triggers 0-5 or the ones before the cancel seem not be called. Seems like cancel_timer cancels unrelated timers as well ? 

Python: 2.7.15
redis==3.2.0

**Comment by tonythomas01** [comment]
Hey @jruizgit, (just observed this happening again on our setup). 

As you can see, `cancel_timer` seems to also expire out other valid (but pending) timers in the queue. If you can point me to where we should be looking at, I can try working on a fix as well.

**Comment by jruizgit** [comment]
Hi, thanks for reporting this issue. The offending code is in src/rules/net.c line 539. The generated lua script is looking for the timer name, however it not taking the sid into account. The fix is a little involved, as we need to pass the sid down to the lua script and look for the right timer.

```c
static unsigned int loadRemoveTimerCommand(ruleset *tree, binding *rulesBinding) {
    char *name = &tree->stringPool[tree->nameOffset];
    redisContext *reContext = rulesBinding->reContext;
    redisReply *reply;
    char *lua = NULL;
    if (asprintf(&lua,
"local timer_key = \"%s!t\"\n"
"local timer_name = ARGV[1]\n"
"local res = redis.call(\"zrange\", timer_key, 0, -1)\n"
"for i = 1, #res, 1 do\n"
"    if string.find(res[i], \"\\\"%%$t\\\"%%s*:%%s*\\\"\" .. timer_name .. \"\\\"\") then\n"
"        redis.call(\"zrem\", timer_key, res[i])\n"
"    end\n"
"end\n"
"return 0\n", name)  == -1) {
        return ERR_OUT_OF_MEMORY;
    }

    unsigned int result = redisAppendCommand(reContext, "SCRIPT LOAD %s", lua);
    GET_REPLY(result, "loadRemoveTimerCommand", reply);

    strncpy(rulesBinding->removeTimerHash, reply->str, 40);
    rulesBinding->removeTimerHash[40] = '\0';
    freeReplyObject(reply);
    free(lua);
    return RULES_OK;
}
```

**Comment by jruizgit** [comment]
I have pushed a fix and uploaded a new version to pypi (0.133.34). With the fix this is the output I get:

```
INFO:root:Starting timer for sid: 0
INFO:root:Starting timer for sid: 1
INFO:root:Starting timer for sid: 2
INFO:root:Starting timer for sid: 3
INFO:root:Starting timer for sid: 4
INFO:root:Starting timer for sid: 5
INFO:root:Cancelling timer for sid: 5
INFO:root:Starting timer for sid: 6
INFO:root:Starting timer for sid: 7
INFO:root:Starting timer for sid: 8
INFO:root:Starting timer for sid: 9
INFO:root:Important timer with sid 0 expired.
INFO:root:Important timer with sid 1 expired.
INFO:root:Important timer with sid 2 expired.
INFO:root:Important timer with sid 3 expired.
INFO:root:Important timer with sid 4 expired.
INFO:root:Important timer with sid 6 expired.
INFO:root:Important timer with sid 7 expired.
INFO:root:Important timer with sid 8 expired.
INFO:root:Important timer with sid 9 expired.
```
Please give it a try and let me know if it works for you.

**Comment by tonythomas01** [comment]
Thank you @jruizgit. 

We deployed `0.33.134` and things seem all fine at this point. Will close this one as resolved and re-open if something shows up. The unit tests seems to all ✅  though.
